src/client.rs

```diff
@@ -0,0 +1,369 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::process;
+
+use std::error::Error;
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+
+use futures::sync::oneshot::Sender;
+use futures::{Future, Poll, Stream};
+
+use tokio::timer::{Delay, Interval};
+
+use uuid::Uuid;
+
+use crate::cmd_handler;
+use crate::net::arrow;
+
+use crate::cmd_handler::{Command, CommandChannel};
+use crate::config::Config;
+use crate::context::{ApplicationContext, ConnectionState};
+use crate::net::arrow::{ArrowError, ErrorKind};
+use crate::net::raw::ether::MacAddr;
+use crate::utils::logger::{BoxLogger, Logger};
+
+/// Connection retry timeout.
+const RETRY_TIMEOUT: f64 = 60.0;
+
+/// Get maximum duration of the pairing mode.
+const PAIRING_MODE_TIMEOUT: f64 = 1200.0;
+
+/// This future ensures maintaining connection with a remote Arrow Service.
+struct ArrowMainTask {
+    app_context: ApplicationContext,
+    cmd_channel: CommandChannel,
+    logger: BoxLogger,
+    default_addr: String,
+    current_addr: String,
+    last_attempt: f64,
+    pairing_mode_timeout: f64,
+    diagnostic_mode: bool,
+}
+
+impl ArrowMainTask {
+    /// Create a new task.
+    #[allow(clippy::new_ret_no_self)]
+    fn new(
+        app_context: ApplicationContext,
+        cmd_channel: CommandChannel,
+    ) -> impl Future<Item = (), Error = ()> {
+        let logger = app_context.get_logger();
+        let addr = app_context.get_arrow_service_address();
+        let diagnostic_mode = app_context.get_diagnostic_mode();
+
+        let t = time::precise_time_s();
+
+        let pairing_mode_timeout = t + PAIRING_MODE_TIMEOUT;
+
+        let task = ArrowMainTask {
+            app_context,
+            cmd_channel,
+            logger,
+            default_addr: addr.clone(),
+            current_addr: addr,
+            last_attempt: t,
+            pairing_mode_timeout,
+            diagnostic_mode,
+        };
+
+        let task = Arc::new(Mutex::new(task));
+
+        let connector = task.clone();
+        let rhandler = task;
+
+        futures::stream::repeat(())
+            .and_then(move |_| connector.lock().unwrap().connect())
+            .then(move |res| rhandler.lock().unwrap().process_result(res))
+            .for_each(|_| Ok(()))
+    }
+
+    /// Connect to the Arrow service.
+    fn connect(&mut self) -> impl Future<Item = String, Error = ArrowError> {
+        log_info!(
+            &mut self.logger,
+            "connecting to remote Arrow Service {}",
+            self.current_addr
+        );
+
+        self.last_attempt = time::precise_time_s();
+
+        self.app_context
+            .set_connection_state(ConnectionState::Connected);
```

Comment:
The content of `src/client.rs` is just a copy-paste from `src/main.rs`. I didn't do any functional changes.

In this particular case, we cannot set "connected" after the future is resolved. The future is pending for the whole time when the connection is active and it gets resolved only when the connection is closed. We also cannot add any new states because it would be a breaking change to the API of the application (the connection state is saved to a file that can be consumed by other applications).

Ideally, we'd set the state within the future implementation but let's do it in a different pull request.